File: runtime/ruby/units/rt/ftrunc.rb
# requires: rt/quiet_nan
def ftrunc(x)
  return quiet_nan(x) if x.nan?
  return x unless x.finite?
  return x if x == 0.0 # preserves -0.0
  r = x.truncate
  r == 0 && x < 0.0 ? -0.0 : r.to_f
end

File: crates/dewasm-backend-python/units/rt/ftrunc.py
# requires: rt/quiet_nan
@staticmethod
def ftrunc(x):
    if x != x:
        return Rt.quiet_nan(x)
    if not math.isfinite(x):
        return x
    if x == 0.0:
        return x
    r = math.trunc(x)
    return -0.0 if (r == 0 and x < 0.0) else float(r)
